### DISCUSSION TITLE
Create a `create-app` like command to generate boilerplate code and directory structure

### DISCUSSION THREAD

**Discussion by p0lygun** [kivy/kivy-github-discussions#3] [Ideas]
This is for `kivy`, but can be made for bulldozer/p4a too. 
The idea is to have a command like `create-project` that does the following

- Create a directory structure (needs to be discussed)
- Subclass App class
- Create base kv file

In my opinion, some of the **benefits** of having this command like this are
- Allows us to have an "official"/"recommended" way of setting up the project
- Makes it easier to dynamically generate CI pipelines as we will have a defined structure
- Much more organized projects and easier to maintain

**Comment by Samael-TLB** [comment]
I agree to the need for such an utility. Internally, our team has also been using such an utility for easiness and consistency. A few thoughts from our experience though:
1) We have experimented with a lot of architectures which are in essence used for UI dev in major production landscapes, these include both desktop, mobile and web. The current best architecture which we found to be the most flexible and suitable with ease of enhancement is what the latest architecture supported by `create-next-app`  is, that is the app router architecture. The recommendations and abstractions are very nice for both small and enterprise grade applications. [Link for understanding project organization strategies](https://nextjs.org/docs/app/building-your-application/routing/colocation#project-organization-strategies).

2) Now regarding the exact project structure that might be employed, I believe that would depend mainly upon the strategy the app crrator/devs want to employ. That is based on common patterns it can be MVP, MVC, MVVM, etc. What we had found from our experience is that all such patterns have an underlying common structure that can be utilized. But, our current approach is to have MVVM, which seems to be Android's recommended way and also subtly recommended from Apple too. But this is the bit, I feel we have to have the utility produce a flexible enough project template for greater flexibility.

3) `KivyMd` also has an utility for project creation and boilerplate generation in [MVC pattern](https://github.com/kivymd/KivyMD/tree/master/kivymd/tools/patterns) that we can also take a look at for inspiration. 

4) In terms of DX in `Kivy`, we have at least two important basic things to address as per my understanding. The first is to have a well prescribed defined pattern for project creation to best practices. The second is to have the dev flow being streamlined more.

5) In detailing out the dev flow being streamlined more, I will highlight the below three main points to address:
    a) The hot reloading with state preservation.
    b) Easy viewing of app live during dev phase ( though `Kivy` supports screen emulation, but for beginners we can may be have a better experience through the utility in current discussion)
    c) Easy build & deployment to emulator/phone ( phone is already present but emulator deployment needs better developer experience[DX]).

6) In relation to the first sub point from point 5 above, we can also explore a global state management library too. But the current internal implementation that we are developing, we are experimenting with having a fine-grained automatic hot reloading that utilizes the module dependency graphs for reloading optimizations and having an internal mechanism for storing of application state and reloading. The second part is to be handled by having the states (assuming them to be `Kivy properties` only) which would all inherit from base property to be patched with a version that initializes a centralized store upon initialization of a new property instance. This can in future be easily extended for a global state management library.

7) Also as an added feature, we can also explore dependency injection too.

Above are just proposals/recommendations based off on our experience. And I'm keen on hearing on the views from others too.

**Comment by misl6** [comment]
> This is for kivy, but can be made for bulldozer/p4a too.

Since Kivy is a full ecosystem, would be nice if this tool prepares the whole environment and scaffolding, for all the development phases.

What I imagine here is a new project that will be responsible for lowering the newcomers' entry barrier, and easing the workflow for us we already know the Kivy ecosystem pretty well, and `create-app` could be a part (and the starting point) of it.

@Samael-TLB already did a pretty analysis here about the key points that this tool should address:
 
> In detailing out the dev flow being streamlined more, I will highlight the below three main points to address:
a) The hot reloading with state preservation.
b) Easy viewing of app live during dev phase ( though Kivy supports screen emulation, but for beginners we can may be have a better experience through the utility in current discussion)
c) Easy build & deployment to emulator/phone ( phone is already present but emulator deployment needs better developer experience[DX]).

`buildozer` already tried to add some of these features at least for Android, but what is missing is a single tool that also takes care of the very first phases.


Some ideas about the possible features:

- `tool-name init`:
    - Creates the directory structure and add the "Hello world" app (and can take an option, so we do not have to force the developer to use a specific structure, but we can suggest some)
    - Install dependencies of the just created project.

- `tool-name update`:
   - If the dependencies have changed, are updated, everywhere.
   - If the `buildozer`, `kivy-ios` and `python-for-android` versions have changed and require a migration of the Android and iOS projects, the migration is processed.

- `tool-name dev`:
    - Run the app in development mode, with hot-reloading enabled. By default runs on the development machine, so there's no need to cross-compile any artifact.
    - Run the app in development mode, on a specific target device, attached to the development machine or an emulator (e.g. via `--target `)
    - Immediately starts the log console.

**Comment by Julian-O** [comment]
Consider using a [cookie-cutter](https://cookiecutter.readthedocs.io/en/stable/) template for this purpose.

**Comment by FilipeMarch** [comment]
Me and a @AccelQuasarDragon have created recently the Kivy Reloader (we were working on this past month). It doesn't have state management yet, but it already hot reloads on the desktop, phone and emulator at the same time. It also has a script for compiling the app, streaming it (with scrcpy) and starting the log console after finishing the compilation. 

This week we're gonna be working on the `create-app` project and we're gonna release the tools considering what you guys said. We hope in the future these tools could be merged into Kivy itself. What is the best way for Kivy Reloader to contribute to the create-app concept?

**Reply by baseplate-admin** [reply]
> Me and a @AccelQuasarDragon have created recently the Kivy Reloader (we were working on this past month). It doesn't have state management yet, but it already hot reloads on the desktop, phone and emulator at the same time. It also has a script for compiling the app, streaming it (with scrcpy) and starting the log console after finishing the compilation.

Wooh that's awesome. But dont you think we have a bit too many hotReloaders?
1. [`kaki`](https://github.com/tito/kaki)
2. [`pythonar`](https://github.com/kivy/pythonar)
3. [`mdhotreloader`](https://github.com/kivymd/KivyMD/tree/1152d3c31229b357336cb9e19cfaa2ba8c6941bc/kivymd/tools/hotreload)

I think it's better to unify everything ( and maybe give you access to the `pythonar` repo on `kivy` )

> This week we're gonna be working on the create-app project and we're gonna release the tools considering what you guys said. We hope in the future these tools could be merged into Kivy itself. What is the best way for Kivy Reloader to contribute to the create-app concept?

I think, it's better to keep this as a package that should not be merged into the `kivy` package. Something like `kivy/reloader` repo on github is better for this.

<sub> on a side note did you folks architecture <code>create-app</code> </sub>

**Reply by FilipeMarch** [reply]
I haven't ever heard about `pythonar` before. It is 6 years old and is not related to Kivy.
Looking at the code of `mdhotreloader`, I don't see any difference from `kaki` code.
Considering this, I think currently there is only a single excellent Kivy hot reload on desktop, which is `kaki`. But we lack a tool with more features. 

Kivy Reloader is focused in reloading multiple devices at the same time, including Android emulator support, and fine-tuning your reload preferences (which files/folders are watched, ignored, and which files are sent to Android) in one configuration file. 

It is highly opiniated, because your Kivy app will necessarily have `async` support with `trio` and will include basic things that Kivy do not implement by default: basic screen navigation, lazy loading screens, handling back buttons, inspector, window control (size, position, always_on_top). Some people may ask "but why should a reloader have an opinion on all these things?", and the reason is that every time I start a new app I need to implement all these things from scratch which are too basic and are present in any reasonable app. I don't think we should "teach" newcomers how to implement screen navigation or lazy load, we simply should say "oh, to go Login Screen please type `change_screen("Login Screen")` and that's it. "To create a new screen execute this script and type the name of the screen".

Today, 31th October, we'll spend at least 3h implementing this new tool (if you wanna participate please send me your Discord nickname); we need this tool, and we will take a look at `cookie-cutter` as suggested by @Julian-O. We are still not sure what the final architecture of `create-app` will be, but one thing is for sure, we'll make our best to make the life of everyone using it easier. We're gonna strongly incentivize people to use `pyenv` and `poetry`. We are taking our inspirations from `react-native` Expo tool. If you have any other tools to suggest us to take a look, please let us know

**Reply by Julian-O** [reply]
> your Kivy app will necessarily have `async` support with `trio`

That's... very opinionated🙂, given the existence of `asyncio` as a batteries-included standard.

**Reply by FilipeMarch** [reply]
`trio` is **much** better... @Cheaterman could explain you why

**Reply by p0lygun** [reply]
so i was going to work on this, should i wait till u guys make your tool public ?

**Reply by FilipeMarch** [reply]
> so i was going to work on this, should i wait till u guys make your tool public ?

@p0lygun yes, I believe you should wait just a little, we hope to publish everything until next week, we are doing our best to reach this point; then if you want you can help us with some issues or some decisions we need to take.

If you want to cooperate with us right now let me know, send me your Discord nickname so we can discuss in real time and pair program and give you access to the repo

**Reply by p0lygun** [reply]
@FilipeMarch sure, i'd love to help, my discord is gala_vs

**Comment by Samael-TLB** [comment]
Any updates happening on this?

**Reply by p0lygun** [reply]
was going to work on it, but @FilipeMarch said they have something already ready for release.. so waiting for that
